crates/relayer-framework/src/tests/relayer_mock/base/impls/chain.rs

```diff
@@ -0,0 +1,171 @@
+use crate::base::one_for_all::traits::runtime::OfaRuntimeContext;
+use crate::common::one_for_all::presets::MinimalPreset;
+use crate::tests::relayer_mock::base::error::Error;
+use crate::tests::relayer_mock::base::types::chain::{ChainStatus, ConsensusState};
+use crate::tests::relayer_mock::base::types::events::{Event, WriteAcknowledgementEvent};
+use crate::tests::relayer_mock::base::types::height::Height;
+use crate::tests::relayer_mock::base::types::message::Message as MockMessage;
+use crate::tests::relayer_mock::base::types::runtime::MockRuntimeContext;
+use crate::tests::relayer_mock::contexts::chain::MockChainContext;
+
+impl OfaChainTypes for MockChainContext {
+    type Preset = MinimalPreset;
+
+    type Error = Error;
+
+    type Runtime = MockRuntimeContext;
+
+    type Height = Height;
+
+    type Timestamp = Height;
+
+    type Message = MockMessage;
+
+    type Signer = u128;
+
+    type RawMessage = String;
+
+    type Event = Event;
+
+    type ClientId = String;
+
+    type ConnectionId = String;
+
+    type ChannelId = String;
+
+    type PortId = String;
+
+    type Sequence = u128;
+
+    type WriteAcknowledgementEvent = WriteAcknowledgementEvent;
+
+    type ConsensusState = ConsensusState;
+
+    type ChainStatus = ChainStatus;
+}
+
+#[async_trait]
+impl OfaBaseChain for MockChainContext {
+    fn runtime(&self) -> &OfaRuntimeContext<MockRuntimeContext> {
+        self.runtime()
+    }
+
+    fn encode_raw_message(
+        message: &Self::Message,
+        _signer: &Self::Signer,
+    ) -> Result<Self::RawMessage, Self::Error> {
+        Ok(message.to_string())
+    }
+
+    // Only single messages are sent by the Mock Chain
+    fn estimate_message_len(_message: &Self::Message) -> Result<usize, Self::Error> {
+        Ok(1)
+    }
+
+    fn chain_status_height(status: &Self::ChainStatus) -> &Self::Height {
+        &status.height
+    }
+
+    fn chain_status_timestamp(status: &Self::ChainStatus) -> &Self::Timestamp {
+        &status.timestamp
+    }
+
+    fn try_extract_write_acknowledgement_event(
+        event: Self::Event,
+    ) -> Option<Self::WriteAcknowledgementEvent> {
+        match event {
+            Event::RecvPacket(_) => Some(WriteAcknowledgementEvent {}),
+            _ => None,
+        }
+    }
+
+    /// If the message is a `SendPacket`, update the received packets,
+    /// and add a `RecvPacket` event to the returned array of events.
+    /// If the message is an `AckPacket`, update the received acknowledgment
+    /// packets.
+    /// If the message is an `UpdateClient` update the consensus state.
+    async fn send_messages(
+        &self,
+        messages: Vec<Self::Message>,
+    ) -> Result<Vec<Vec<Self::Event>>, Error> {
+        let mut res = vec![];
+        for m in messages {
+            match m {
+                MockMessage::SendPacket(_, _, h, p) => {
+                    self.receive_packet(p)?;
```

Comment:
We should check on the consensus state on whether there is a consensus state at the _exact_ height given. If so, we should also check if the consensus state contains the packet in the list of sent packets. If not, it should return an error.

The same applies to `AckPacket`.